protocol/packages/currencies/build/main.rs

```diff
@@ -0,0 +1,148 @@
+        .map(Path::try_exists)
+        .try_fold(true, |all_exist, result| {
+            result
+                .map(|exists| all_exist && exists)
+                .context("Failed to check whether JSON descriptor file exists!")
+        })?;
+
+    let build_report = if let Some(report_file) = env::var_os(BUILD_REPORT) {
+        if files_exist {
+            Either::Left(
+                File::create(report_file).context("Failed to open build report for writing!")?,
+            )
+        } else {
+            bail!(
+                "`{BUILD_REPORT:?}` environment variable set but topology \
+                and/or protocol descriptors don't exist!",
+            );
+        }
+    } else {
+        Either::Right(io::stderr())
+    };
+
+    if files_exist {
+        generate_currencies(
+            build_report,
+            output_directory,
+            serde_json::from_reader(
+                File::open(TOPOLOGY_JSON).context("Failed to open \"topology.json\"!")?,
+            )
+            .context("Failed to parse topology JSON!")?,
+            serde_json::from_reader(
+                File::open(PROTOCOL_JSON).context("Failed to open \"protocol.json\"!")?,
+            )
+            .context("Failed to parse protocol JSON!")?,
+        )
+    } else {
+        Ok(())
+    }
+}
+
+fn generate_currencies<BuildReport>(
+    build_report: BuildReport,
+    output_directory: &Path,
+    topology: Topology,
+    protocol: Protocol,
+) -> Result<()>
+where
+    BuildReport: Write,
+{
+    let CurrencyDefinitions {
+        host_currency,
+        dex_currencies,
+    } = topology.currency_definitions(&protocol.dex_network)?;
+
+    if *protocol.lpn_ticker == *host_currency.ticker() {
+        bail!(
+            "Liquidity provider's currency cannot be the same as the host \
+                network's native currency!",
+        );
+    }
+
+    if *protocol.stable_currency_ticker == *host_currency.ticker() {
+        bail!(
+            "Stable currency cannot be the same as the host network's native \
+                currency!",
+        );
+    }
+
+    let dex_currencies = dex_currencies
+        .iter()
+        .filter(|currency_definition| {
+            filter_selected_currencies(
+                &protocol,
+                host_currency.ticker(),
+                currency_definition.ticker(),
+            )
+        })
+        .map(|currency_definition| {
+            (
+                currency_definition.ticker(),
+                (
+                    convert_case::snake_case_to_upper_camel_case(currency_definition.ticker()),
+                    currency_definition,
+                ),
+            )
+        })
+        .collect();
+
+    sources::write(
+        build_report,
+        output_directory,
+        &protocol,
+        &host_currency,
+        &dex_currencies,
+        &CurrenciesTree::new(&topology, &protocol, host_currency.ticker())?,
+    )
+}
+
+#[inline]
+fn filter_selected_currencies(
```

Comment:
name it better - for example `protocol_currency`
and make it a member of `Protocol`